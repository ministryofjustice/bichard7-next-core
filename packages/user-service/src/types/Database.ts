import { IDatabase } from "pg-promise"
import pg from "pg-promise/typescript/pg-subset"

// eslint-disable-next-line @typescript-eslint/ban-types
type Database = IDatabase<{}, pg.IClient>

export default Database
