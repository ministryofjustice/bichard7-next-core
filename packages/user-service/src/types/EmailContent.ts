export default interface EmailContent {
  subject: string
  text: string
  html?: string
}
