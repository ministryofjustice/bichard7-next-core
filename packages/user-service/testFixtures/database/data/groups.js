const groups = [
  {
    name: "B7Allocator_grp",
    friendly_name: "Allocator"
  },
  {
    name: "B7Audit_grp",
    friendly_name: "Audit"
  },
  {
    name: "B7ExceptionHandler_grp",
    friendly_name: "Exception Handler"
  },
  {
    name: "B7GeneralHandler_grp",
    friendly_name: "General Handler"
  },
  {
    name: "B7Supervisor_grp",
    friendly_name: "Supervisor"
  },
  {
    name: "B7TriggerHandler_grp",
    friendly_name: "Trigger Handler"
  },
  {
    name: "B7UserManager_grp",
    friendly_name: "User Manager"
  },
  {
    name: "B7AuditLoggingManager_grp",
    friendly_name: ""
  },
  {
    name: "B7SuperUserManager_grp",
    friendly_name: "Super User Manager"
  },
  {
    name: "B7NewUI_grp",
    friendly_name: "New UI"
  }
]

export default groups
