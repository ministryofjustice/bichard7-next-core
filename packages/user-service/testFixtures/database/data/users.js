const users = [
  {
    username: "Bichard01",
    created_at: "2021-08-01 00:00:00-00",
    endorsed_by: "endorsed_by 01",
    last_logged_in: "2021-08-01 00:00:00-00",
    org_serves: "org_severs 01",
    forenames: "Bichard User 01",
    surname: "Surname 01",
    email: "bichard01@example.com",
    password:
      "$argon2id$v=19$m=256,t=20,p=2$TTFCN3BRcldZVUtGejQ3WE45TGFqPT0$WOE+jDILDnVIAt1dytb+h65uegrMomp2xb0Q6TxbkLA",
    last_login_attempt: "2021-08-01 00:00:00-00",
    deleted_at: null,
    password_reset_code: null,
    migrated_password: null,
    visible_forces: "001,002,004,014",
    visible_courts: "B01,B41ME00",
    excluded_triggers: "TRPR0001",
    feature_flags: { httpsRedirect: true }
  },
  {
    username: "Bichard02",
    created_at: "2021-08-01 00:00:00-00",
    endorsed_by: "endorsed_by 02",
    last_logged_in: "2021-08-01 00:00:00-00",
    org_serves: "org_severs 02",
    forenames: "Bichard User 02",
    surname: "Surname 02",
    email: "bichard02@example.com",
    password:
      "$argon2id$v=19$m=256,t=20,p=2$TTFCN3BRcldZVUtGejQ3WE45TGFqPT0$WOE+jDILDnVIAt1dytb+h65uegrMomp2xb0Q6TxbkLA",
    last_login_attempt: "2021-08-01 00:00:00-00",
    deleted_at: null,
    password_reset_code: null,
    migrated_password: null,
    visible_forces: "001,002,004",
    visible_courts: "B41ME00",
    excluded_triggers: "TRPR0004",
    feature_flags: { httpsRedirect: true }
  },
  {
    username: "Bichard03",
    created_at: "2021-08-01 00:00:00-00",
    endorsed_by: "endorsed_by 03",
    last_logged_in: "2021-08-01 00:00:00-00",
    org_serves: "org_severs 03",
    forenames: "Bichard User 03",
    surname: "Surname 03",
    email: "bichard03@example.com",
    password:
      "$argon2id$v=19$m=256,t=20,p=2$TTFCN3BRcldZVUtGejQ3WE45TGFqPT0$WOE+jDILDnVIAt1dytb+h65uegrMomp2xb0Q6TxbkLA",
    last_login_attempt: "2021-08-01 00:00:00-00",
    deleted_at: null,
    password_reset_code: null,
    migrated_password: null,
    visible_forces: "014",
    visible_courts: "",
    excluded_triggers: ""
  },
  {
    username: "Bichard04",
    created_at: "2021-08-01 00:00:00-00",
    endorsed_by: "endorsed_by 04",
    last_logged_in: "2021-08-01 00:00:00-00",
    org_serves: "org_severs 04",
    forenames: "Bichard User 04",
    surname: "Surname 04",
    email: "bichard04@example.com",
    password:
      "$argon2id$v=19$m=256,t=20,p=2$TTFCN3BRcldZVUtGejQ3WE45TGFqPT0$WOE+jDILDnVIAt1dytb+h65uegrMomp2xb0Q6TxbkLA",
    last_login_attempt: "2021-08-01 00:00:00-00",
    deleted_at: null,
    password_reset_code: null,
    migrated_password: null,
    visible_forces: "",
    visible_courts: "",
    excluded_triggers: ""
  },
  {
    username: "DeletedUser",
    created_at: "2021-08-01 00:00:00-00",
    endorsed_by: "endorsed_by 03",
    last_logged_in: "2021-08-01 00:00:00-00",
    org_serves: "org_severs 03",
    forenames: "Bichard User 03",
    surname: "Surname 03",
    email: "deleted@example.com",
    password:
      "$argon2id$v=19$m=256,t=20,p=2$TTFCN3BRcldZVUtGejQ3WE45TGFqPT0$WOE+jDILDnVIAt1dytb+h65uegrMomp2xb0Q6TxbkLA",
    last_login_attempt: "2021-08-01 00:00:00-00",
    deleted_at: "2021-08-01 00:00:00-00",
    password_reset_code: null,
    migrated_password: null,
    visible_forces: "014",
    visible_courts: "",
    excluded_triggers: ""
  }
]

export default users
