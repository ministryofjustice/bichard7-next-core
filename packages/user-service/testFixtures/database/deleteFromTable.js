import getTestConnection from "../getTestConnection"
import { getTableName } from "./helpers"

const deleteFromTable = async (tableName, whereColumn, whereValue) => {
  const connection = getTestConnection()
  const isWhereClause = whereColumn && whereValue

  const deleteQuery = `DELETE FROM br7own.users_groups; DELETE FROM $\{table\} ${
    isWhereClause ? `WHERE ${whereColumn} = $\{value\}` : ""
  }`

  const table = getTableName(tableName)
  const whereClause = isWhereClause ? { column: whereColumn, value: whereValue } : {}

  return connection.none(deleteQuery, { table, ...whereClause })
}

export default deleteFromTable
