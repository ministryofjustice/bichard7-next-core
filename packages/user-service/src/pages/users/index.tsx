import Button from "components/Button"
import ButtonGroup from "components/ButtonGroup"
import Form from "components/Form"
import Layout from "components/Layout"
import Link from "components/Link"
import SuccessBanner from "components/SuccessBanner"
import { LinkColumn, Table, TableHeaders } from "components/Table"
import TextInput from "components/TextInput"
import config from "lib/config"
import getConnection from "lib/getConnection"
import { withAuthentication, withCsrf, withMultipleServerSideProps } from "middleware"
import { GetServerSidePropsContext, GetServerSidePropsResult } from "next"
import Head from "next/head"
import { ParsedUrlQuery } from "querystring"
import useCustomStyles from "styles/useCustomStyles"
import AuthenticationServerSidePropsContext from "types/AuthenticationServerSidePropsContext"
import CsrfServerSidePropsContext from "types/CsrfServerSidePropsContext"
import KeyValuePair from "types/KeyValuePair"
import { isError } from "types/Result"
import User from "types/User"
import getAllUsers from "useCases/getAllUsers"
import getFilteredUsers from "useCases/getFilteredUsers"
import getUserServiceAccess, { type UserServiceAccess } from "useCases/getUserServiceAccess"
import isUserWithinGroup from "useCases/isUserWithinGroup"
import addQueryParams from "utils/addQueryParams"
import createRedirectResponse from "utils/createRedirectResponse"
import { isPost } from "utils/http"
import logger from "utils/logger"

export const getServerSideProps = withMultipleServerSideProps(
  withAuthentication,
  withCsrf,
  async (context: GetServerSidePropsContext<ParsedUrlQuery>): Promise<GetServerSidePropsResult<Props>> => {
    const { req, query, formData, csrfToken, currentUser, authentication } = context as CsrfServerSidePropsContext &
      AuthenticationServerSidePropsContext
    const connection = getConnection()
    let pageNumber = 0
    let previousFilter = ""
    let bannerMessage = ""

    if (!currentUser || !authentication) {
      return createRedirectResponse("/login")
    }

    const hasAccessTo = getUserServiceAccess(authentication)

    if (isPost(req)) {
      const { filter } = formData as {
        filter: string
      }
      previousFilter = filter
    } else {
      const { filter, page, action } = query as {
        filter: string
        page: string
        action: string
      }

      if (filter) {
        previousFilter = filter
      }

      if (page) {
        pageNumber = parseInt(page, 10)
      }

      switch (action) {
        case "user-created":
          bannerMessage = "User created successfully."
          break
        case "user-deleted":
          bannerMessage = "User deleted successfully."
          break
        default:
      }
    }

    const isCurrentSuperUser = await isUserWithinGroup(connection, currentUser?.id || -1, "B7SuperUserManager")

    let queryResult
    if (previousFilter) {
      queryResult = await getFilteredUsers(
        connection,
        previousFilter,
        currentUser?.visibleForces ?? "",
        isCurrentSuperUser,
        pageNumber
      )
    } else {
      queryResult = await getAllUsers(connection, currentUser?.visibleForces ?? "", isCurrentSuperUser, pageNumber)
    }

    if (isError(queryResult)) {
      logger.error(queryResult)
      return {
        props: {
          allUsers: null,
          csrfToken,
          currentUser,
          previousFilter,
          pageNumber,
          totalUsers: 0,
          bannerMessage,
          hasAccessTo
        }
      }
    }

    const { result, totalElements } = queryResult
    const allUsers = result
    const totalUsers = totalElements

    return {
      props: {
        allUsers: allUsers as KeyValuePair<string, string>[] | null,
        csrfToken,
        currentUser,
        previousFilter,
        pageNumber,
        totalUsers,
        bannerMessage,
        hasAccessTo
      }
    }
  }
)

interface Props {
  allUsers: KeyValuePair<string, string>[] | null
  csrfToken: string
  currentUser?: Partial<User>
  previousFilter: string
  pageNumber: number
  totalUsers: number
  bannerMessage?: string
  hasAccessTo: UserServiceAccess
}

const tableHeaders: TableHeaders = [
  ["username", "Username"],
  ["forenames", "Forename(s)"],
  ["surname", "Surname"],
  ["emailAddress", "Email address"]
]

const Users = ({
  allUsers,
  csrfToken,
  currentUser,
  previousFilter,
  pageNumber,
  totalUsers,
  bannerMessage,
  hasAccessTo
}: Props) => {
  const nextPage = addQueryParams("/users", {
    filter: previousFilter,
    page: pageNumber + 1
  })

  const prevPage = addQueryParams("/users", {
    filter: previousFilter,
    page: pageNumber - 1
  })

  const pageNumberStyle = {
    style: {
      padding: "10px"
    },
    spacing: {
      paddingLeft: "0px"
    }
  }

  let styles = { ...pageNumberStyle.style }

  if (pageNumber === 0) {
    styles = { ...styles, ...pageNumberStyle.spacing }
  }

  const pageString = `Page ${pageNumber + 1} of ${Math.ceil(totalUsers / config.maxUsersPerPage)}`
  const classes = useCustomStyles()

  return (
    <>
      <Head>
        <title>{"Users"}</title>
      </Head>
      <Layout user={currentUser} hasAccessTo={hasAccessTo}>
        <div className={`${classes["top-padding"]}`}>
          <h1 className="govuk-heading-l">{"Users"}</h1>

          {!!bannerMessage && <SuccessBanner>{bannerMessage}</SuccessBanner>}

          <Form method="post" csrfToken={csrfToken}>
            <ButtonGroup>
              <Link id="add" className="govuk-button govuk-!-margin-right-8" href="/users/new-user">
                {"Add user"}
              </Link>
              <TextInput className="align-right" id="filter" name="filter" type="text" value={previousFilter} />
              <Button className="govuk-!-margin-left-4" noDoubleClick id="filter">
                {"Filter"}
              </Button>
            </ButtonGroup>
          </Form>

          {allUsers && (
            <Table tableHeaders={tableHeaders} tableData={allUsers}>
              <LinkColumn
                data-test="link-to-user-view"
                field="username"
                href={(user) => `users/${(user as unknown as User).username}`}
              />
            </Table>
          )}

          <div className="govuk-hint">
            <Link href={prevPage} data-test="Prev">
              {pageNumber > 0 && "< Prev"}
            </Link>
            <span style={styles}>{pageString}</span>
            <Link href={nextPage} data-test="Next">
              {pageNumber + 1 <= (totalUsers - 1) / config.maxUsersPerPage && "Next >"}
            </Link>
          </div>
        </div>
      </Layout>
    </>
  )
}

export default Users
