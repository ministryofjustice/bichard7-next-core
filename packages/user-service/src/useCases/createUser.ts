import { ITask } from "pg-promise"
import Database from "types/Database"
import PromiseResult from "types/PromiseResult"
import { isError } from "types/Result"
import User from "types/User"
import { UserGroupResult } from "types/UserGroup"
import logger from "utils/logger"
import isEmailUnique from "./IsEmailUnique"
import getUserHierarchyGroups from "./getUserHierarchyGroups"
import isUsernameUnique from "./isUsernameUnique"

type InsertUserResult = PromiseResult<{ id: number }>

const insertUser = (task: ITask<unknown>, userDetails: Partial<User>): InsertUserResult => {
  const insertUserQuery = `
      INSERT INTO br7own.users(
        username,
        forenames,
        surname,
        email,
        endorsed_by,
        org_serves,
        visible_courts,
        visible_forces,
        excluded_triggers,
        feature_flags
      )
      VALUES (
        $\{username\},
        $\{forenames\},
        $\{surname\},
        $\{emailAddress\},
        $\{endorsedBy\},
        $\{orgServes\},
        $\{visibleCourts\},
        $\{visibleForces\},
        $\{excludedTriggers\},
        $\{featureFlags\}
      ) RETURNING id;
    `

  return task.one(insertUserQuery, { ...userDetails }).catch((error) => error)
}

const insertUserIntoGroup = async (
  task: ITask<unknown>,
  newUserId: number,
  groups: UserGroupResult[]
): PromiseResult<void> => {
  const groupIds = groups.map((group: UserGroupResult) => group.id)

  const insertGroupQuery = `
    INSERT INTO br7own.users_groups(
      user_id,
      group_id
    )
    SELECT 
      $\{newUserId\} AS user_id,
      g.id
    FROM br7own.groups AS g
      WHERE
        g.id IN ($\{groupIds:csv\});
  `

  const result = await task.result(insertGroupQuery, { newUserId, groupIds }).catch((error) => error as Error)

  if (isError(result)) {
    logger.error(result)
    return result
  }

  return undefined
}

export default async (
  connection: Database,
  currentUser: Partial<User>,
  userDetails: Partial<User>
): PromiseResult<void> => {
  const tmpUserDetails: Partial<User> = {}
  for (const [key, value] of Object.entries(userDetails)) {
    if (typeof value === "string") {
      tmpUserDetails[key] = String(value).trim()
    } else {
      tmpUserDetails[key] = value
    }
  }
  userDetails = tmpUserDetails

  const isUsernameUniqueResult = await isUsernameUnique(connection, userDetails.username as string)
  if (isError(isUsernameUniqueResult)) {
    return isUsernameUniqueResult
  }

  const isEmailUniqueResult = await isEmailUnique(connection, userDetails.emailAddress as string)
  if (isError(isEmailUniqueResult)) {
    return isEmailUniqueResult
  }

  const createUserResult = await connection
    .tx(async (task: ITask<unknown>) => {
      const insertUserResult = await insertUser(task, userDetails)
      if (isError(insertUserResult)) {
        logger.error(insertUserResult)
        return Error("Could not insert record into users table")
      }
      const groups = await getUserHierarchyGroups(connection, currentUser.username ?? "")
      if (isError(groups)) {
        return groups
      }
      const filteredGroups = groups.filter((group) => userDetails[group.name] === "yes")
      const userDetailsWithGroups = { ...userDetails, groups: filteredGroups }
      if (currentUser.id && userDetailsWithGroups.groups && userDetailsWithGroups.groups.length > 0) {
        const userGroupResult = await insertUserIntoGroup(task, insertUserResult.id, userDetailsWithGroups.groups)
        return userGroupResult
      }

      return undefined
    })
    .catch((error) => error)

  if (isError(createUserResult)) {
    logger.error(createUserResult)
    return Error("Could not create user")
  }

  return undefined
}
