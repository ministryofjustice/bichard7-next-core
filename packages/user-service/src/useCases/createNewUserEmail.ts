import generateNewUserEmail from "emails/newUser"
import { addBasePath } from "next/dist/client/add-base-path"
import EmailContent from "types/EmailContent"
import { Result } from "types/Result"
import User from "types/User"

export default (user: User, baseUrl: string): Result<EmailContent> => {
  const url = new URL(addBasePath("/login/reset-password"), baseUrl)

  return generateNewUserEmail({ url: url.href, user })
}
