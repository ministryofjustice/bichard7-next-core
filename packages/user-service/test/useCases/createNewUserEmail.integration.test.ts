/**
 * @jest-environment node
 */

import Database from "types/Database"
import EmailContent from "types/EmailContent"
import { isError } from "types/Result"
import createNewUserEmail from "useCases/createNewUserEmail"
import createUser from "useCases/createUser"
import groups from "../../testFixtures/database/data/groups"
import users from "../../testFixtures/database/data/users"
import deleteFromTable from "../../testFixtures/database/deleteFromTable"
import insertIntoGroupsTable from "../../testFixtures/database/insertIntoGroupsTable"
import insertIntoUserGroupsTable from "../../testFixtures/database/insertIntoUserGroupsTable"
import insertIntoUsersTable from "../../testFixtures/database/insertIntoUsersTable"
import selectFromTable from "../../testFixtures/database/selectFromTable"
import getTestConnection from "../../testFixtures/getTestConnection"

describe("AccountSetup", () => {
  let connection: Database

  beforeAll(() => {
    connection = getTestConnection()
  })

  beforeEach(async () => {
    await deleteFromTable("users_groups")
    await deleteFromTable("users")
    await deleteFromTable("groups")
  })

  afterAll(() => {
    connection.$pool.end()
  })

  it("should generate the email subject and body to request user to setup password", async () => {
    const username = "Bichard02"
    await insertIntoUsersTable(users.filter((u) => u.username !== username))
    await insertIntoGroupsTable(groups)
    await insertIntoUserGroupsTable(
      "bichard01@example.com",
      groups.map((g) => g.name)
    )
    const currentUserId = (await selectFromTable("users", "username", "Bichard01"))[0].id

    const selectedGroups = await selectFromTable("groups", undefined, undefined, "name")
    const selectedGroup = selectedGroups[0]

    const user = users
      .filter((u) => u.username === "Bichard02")
      .map((u) => ({
        username: u.username,
        forenames: u.forenames,
        emailAddress: u.email,
        endorsedBy: u.endorsed_by,
        surname: u.surname,
        orgServes: u.org_serves,
        groups: [selectedGroup],
        visibleForces: "001,004,",
        visibleCourts: "B01,B41ME00",
        excludedTriggers: "TRPR0001",
        featureFlags: { httpsRedirect: true }
      }))[0]

    const result = await createUser(connection, { id: currentUserId, username: "Bichard01" }, user)

    expect(isError(result)).toBe(false)

    const newUserEmailResult = createNewUserEmail(user, "http://localhost:3000")
    expect(isError(newUserEmailResult)).toBe(false)

    const email = newUserEmailResult as EmailContent
    expect(email.subject).toMatchSnapshot()
    expect(email.text).toMatchSnapshot()
    expect(email.html).toMatchSnapshot()
  })
})
