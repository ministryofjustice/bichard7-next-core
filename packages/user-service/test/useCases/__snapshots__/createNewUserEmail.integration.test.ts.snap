// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`AccountSetup should generate the email subject and body to request user to setup password 1`] = `"Finish Bichard account setup"`;

exports[`AccountSetup should generate the email subject and body to request user to setup password 2`] = `
"Hi Bichard User 02 Surname 02,

In order to finish setting up your Bichard account, you must create a password. You can do this by clicking the link below.

http://localhost:3000/login/reset-password
"
`;

exports[`AccountSetup should generate the email subject and body to request user to setup password 3`] = `"<html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/><title>Finish setting up your Bichard account</title></head><body><table style="background-color:#f3f2f1;margin:0;padding:30px;width:100%"><tr></tr><tr><td style="background-color:#ffffff;border:1px solid;border-color:#b1b4b6;clear:both;display:block !important;margin:0 auto;max-width:600px !important;padding:30px;vertical-align:top" valign="top" width="600"><h1 style="color:#0b0c0c;font-family:Arial, sans-serif;font-size:48px;line-height:50px;margin-bottom:50px">Finish setting up your Bichard account</h1><p style="color:#0b0c0c;font-family:Arial, sans-serif;font-size:19px;line-height:25px;margin-bottom:20px">Hi Bichard User 02 Surname 02,</p><p style="color:#0b0c0c;font-family:Arial, sans-serif;font-size:19px;line-height:25px;margin-bottom:20px">In order to finish setting up your Bichard account, you need to verify your email address and create a password. You can do this by clicking the button below.</p><a href="http://localhost:3000/login/reset-password" style="background-color:#00703c;color:#ffffff;cursor:pointer;display:inline-block;font-family:Arial, sans-serif;font-size:19px;line-height:19px;padding:8px 10px 7px 10px;text-align:center;text-decoration:none">Create password</a><p style="color:#505a5f;font-family:Arial, sans-serif;font-size:14px;line-height:16px;margin-bottom:16px;margin-top:50px">If youâ€™re having trouble clicking the button, copy and paste the URL below into your web browser:</p><p style="color:#505a5f;font-family:Arial, sans-serif;font-size:14px;line-height:16px;margin-bottom:16px;margin:0"><a href="http://localhost:3000/login/reset-password" style="color:#505a5f;font-family:Arial, sans-serif;font-size:14px;line-height:16px;margin-bottom:16px">http://localhost:3000/login/reset-password</a></p></td></tr><tr></tr></table></body></html>"`;
